src/include/catalog/postgres/pg_proc_impl.h

```diff
@@ -0,0 +1,178 @@
+}  // namespace noisepage::execution::functions
+
+namespace noisepage::storage {
+class RecoveryManager;
+class SqlTable;
+
+namespace index {
+class Index;
+}  // namespace index
+}  // namespace noisepage::storage
+
+namespace noisepage::transaction {
+class TransactionContext;
+}  // namespace noisepage::transaction
+
+namespace noisepage::catalog::postgres {
+class Builder;
+
+/** The NoisePage version of pg_proc. */
+class PgProcImpl {
+ public:
+  /**
+   * Prepare to create pg_proc.
+   * Does NOT create anything until the relevant bootstrap functions are called.
+   *
+   * @param db_oid          The OID of the database that pg_proc should be created in.
+   */
+  explicit PgProcImpl(db_oid_t db_oid);
+
+  /** Bootstrap the projected row initializers for pg_proc. */
+  void BootstrapPRIs();
+
+  /**
+   * Bootstrap:
+   *    pg_proc
+   *    pg_proc_oid_index
+   *    pg_proc_name_index
+   *
+   * @param dbc             The catalog object to bootstrap in.
+   * @param txn             The transaction to bootstrap in.
+   */
+  void Bootstrap(common::ManagedPointer<DatabaseCatalog> dbc,
+                 common::ManagedPointer<transaction::TransactionContext> txn);
+
+  /**
+   * Get all of the function contexts, used in TearDown.
+   * A buffer is passed in to allow for buffer reuse optimizations.
+   *
+   * @param txn             The transaction to obtain all the function contexts in.
+   * @param buffer          The buffer to use for getting function contexts back out.
+   * @param buffer_len      The length of buffer. Only used in debug mode to assert correctness.
+   * @return All the function contexts.
+   */
+  std::vector<execution::functions::FunctionContext *> TearDownGetFuncContexts(
+      common::ManagedPointer<transaction::TransactionContext> txn, byte *buffer, UNUSED_ATTRIBUTE uint64_t buffer_len);
+
+  /**
+   * Create a procedure for the pg_proc table.
+   *
+   * @param txn             The transaction to use.
+   * @param oid             The OID to assign to the procedure.
+   * @param procname        The name of the procedure.
+   * @param language_oid    The OID for the language that this procedure is written in.
+   * @param procns          The namespace that the procedure should be added to.
+   * @param args            The names of the arguments to this procedure.
+   * @param arg_types       The types of the arguments to this procedure. Must be in the same order as in args.
+   *                        (only for in and inout arguments)
+   * @param all_arg_types   The types of all the arguments.
+   * @param arg_modes       The modes of the arguments. Must be in the same order as in args.
+   * @param rettype         The OID of the type of return value.
+   * @param src             The source code of the procedure.
+   * @param is_aggregate    True iff this is an aggregate procedure.
+   * @return True if the creation succeeded. False otherwise.
+   *
+   * @warning Does not support variadics yet.
```

Comment:
This wouldn't be possible unless I add `DatabaseCatalog` as an argument, since `GetTypeOidForType` is a `DatabaseCatalog` thing.

Reply:
Thoughts?

Reply:
I guess I don't see that as adding an arg just for debugging.  We're going to need it eventually to support the feature, so it's implementing the necessary API even if we don't use it yet.  @mbutrovich, do you have thoughts on this one?

Reply:
We spoke briefly. If/when the refactor happens, we can modify it accordingly. For now, we push the assert one level up -- technically incorrect, but we have access to the DBC there and it would throw the assert with the right error message.

Reply:
Actually.. we create concat which is variadic. Perhaps this is just an outdated warning.

Reply:
I removed the warning.. fingers crossed.